File: web/js/math.js
export class Vec2 {
    constructor(x, y) {
        this.x = x;
        this.y = y;
    }

    array() {
        return [this.x, this.y];
    }

    scale(s) {
        return new Vec2(s * this.x, s * this.y);
    }

    dot(other) {
        return this.x * other.x + this.y * other.y;
    }

    add(other) {
        return new Vec2(this.x + other.x, this.y + other.y);
    }

    subtract(other) {
        return new Vec2(this.x - other.x, this.y - other.y);
    }

    rotate(angle) {
        const s = Math.sin(angle);
        const c = Math.cos(angle);
        const x = c * this.x + s * this.y;
        const y = -s * this.x + c * this.y;
        return new Vec2(x, y);
    }

    orth() {
        return new Vec2(this.y, -this.x);
    }

    length() {
        return Math.sqrt(this.dot(this));
    }

    unit() {
        const d = this.length();
        if (d > 0) {
            return this.scale(1 / d);
        }
        return this;
    }
}


export function angle2pi(v, start=0) {
    // negative for y is because we are in a left-handed frame
    let a = Math.atan2(-v.y, v.x) - start;
    if (a < 0) {
        a += 2 * Math.PI;
    }
    return a;
}


export function wrapToPi(x) {
    // Wrap a value to [-pi, pi]
    while (x > Math.PI) {
        x -= 2 * Math.PI;
    }
    while (x < -Math.PI) {
        x += 2 * Math.PI;
    }
    return x;
}

File: shadows/math.py
import math
import numpy as np

ORTHMAT = np.array([[0, 1], [-1, 0]])

def unit(v):
    """Normalize to a unit vector."""
    norm = np.linalg.norm(v)
    if np.isclose(norm, 0):
        return np.zeros_like(v)
    return v / norm


def orth(v):
    """Generate a 2D orthogonal to v."""
    return np.array([v[1], -v[0]])


def rotmat(angle):
    """2D rotation matrix."""
    c = np.cos(angle)
    s = np.sin(angle)
    return np.array([[c, s], [-s, c]])


def wrap_to_pi(x):
    """Wrap a value to [-pi, pi]"""
    return math.remainder(x, 2 * np.pi)


def angle2pi(v, start=0):
    """Compute angle of vector v w.r.t. the start in the interval [0, 2pi]."""
    # negative for y is because we are in a left-handed frame
    a = np.arctan2(-v[1], v[0]) - start
    if a < 0:
        a = 2 * np.pi + a
    return a


def quad_formula(a, b, c):
    """Evaluate the quadratic formula for coefficients a, b, c.

    The two solutions are returned.
    """
    d = np.sqrt(b ** 2 - 4 * a * c)
    return (-b - d) / (2 * a), (-b + d) / (2 * a)

